llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp

```diff
@@ -2131,36 +2141,121 @@ void DwarfDebug::beginInstruction(const MachineInstr *MI) {
+      // frame setup, and a compiler-generated line 0 location is not a
+      // meaningful breakpoint. If none is found, return the first
+      // location after the frame setup.
+      if (MI.getDebugLoc().getLine())
+        return std::make_pair(&MI, IsEmptyPrologue);
+
+      LineZeroLoc = &MI;
+    }
+
+    // Keep track of the first "non-trivial" instruction seen, i.e. anything
+    // that doesn't involve shuffling data around or is a frame-setup.
+    if (!isCopy && !isTrivRemat && !isFrameSetup && !NonTrivialInst)
+      NonTrivialInst = &MI;
+
+    IsEmptyPrologue = false;
+    return std::nullopt;
+  };
+
+  // Examine all the instructions at the start of the function. This doesn't
+  // necessarily mean just the entry block: unoptimised code can fall-through
+  // into an initial loop, and it makes sense to put the initial breakpoint on
+  // the first instruction of such a loop. However, if we pass branches, we're
+  // better off synthesising an early prologue_end.
+  auto CurBlock = MF->begin();
+  auto CurInst = CurBlock->begin();
+  while (true) {
+    // Skip empty blocks, in rare cases the entry can be empty.
+    if (CurInst == CurBlock->end()) {
+      ++CurBlock;
+      CurInst = CurBlock->begin();
+      continue;
+    }
+
+    // Check whether this non-meta instruction a good position for prologue_end.
+    if (!CurInst->isMetaInstruction()) {
+      auto FoundInst = ExamineInst(*CurInst);
+      if (FoundInst)
+        return *FoundInst;
+    }
+
+    // Try to continue searching, but use a backup-location if substantive
+    // computation is happening.
+    auto NextInst = std::next(CurInst);
+    if (NextInst == CurInst->getParent()->end()) {
+      // We've reached the end of the block. Did we just look at a terminator?
+      if (CurInst->isTerminator())
+        // Some kind of "real" control flow is occurring. At the very least
+        // we would have to start exploring the CFG, a good signal that the
+        // prologue is over.
+        break;
+
+      // If we've already fallen through into a loop, don't fall through
+      // further, use a backup-location.
+      if (CurBlock->pred_size() > 1)
+        break;
+
+      // Fall-through from entry to the next block. This is common at -O0 when
+      // there's no initialisation in the function.
+      auto NextBBIter = std::next(CurInst->getParent()->getIterator());
+      // Bail if we're also at the end of the function.
+      if (NextBBIter == MF->end())
+        break;
+      CurBlock = NextBBIter;
+      CurInst = NextBBIter->begin();
+    } else {
+      // Continue examining the current block.
+      CurInst = NextInst;
     }
   }
-  return std::make_pair(LineZeroLoc, IsEmptyPrologue);
+
+  // We didn't find a "good" prologue_end, so consider backup locations.
+  // Was there an empty-line location? Return that, and it'll have the
+  // scope-line "flow" into it when it becomes the prologue_end position.
```

Comment:
I think this could be slightly more clear, something like: "If we found a line-zero location, use that. The prologue_end entry always gets assigned the scope line."


Wait, but maybe I'm wrong - if `IsEmptyPrologue` is true then wouldn't the line zero get used (I'm looking in `emitInitialLocDirective` - non-null line-zero `PrologEndLoc` and with `IsEmptyPrologue` true)? Is this case handled correctly by your patch?

Reply:
(Hoisting to outer discussion as github has lost this review comment in the diff-display),